File: 2023/Day2/day2.py
file1 = open('input.txt', 'r')
Lines = [ line.strip('\n') for line in file1.readlines() ]

# Part 1
def part_1(Games):
    impossibles = []
    possibles=[]
    maxB= 14
    maxG= 13
    maxR= 12
    for game in Games:
        impossible = False
        gameId=int(game.split(":")[0].split("Game")[1].strip() )
        draws =game.split(":")[1].split(';')
        for draw in draws:
            red = blue = green = 0
            #print(draw)
            for colour in draw.split(","):
                if 'blue' in colour:
                    blue = int(colour.split('blue')[0].strip())
                    if blue > maxB: impossible = True
                if 'red' in colour:
                    red = int(colour.split('red')[0].strip())
                    if red > maxR: impossible = True
                if 'green' in colour:
                    green = int(colour.split('green')[0].strip())
                    if green > maxG: impossible = True
        if impossible: impossibles.append(gameId)
        else: possibles.append(gameId)
    return sum(possibles)

print(part_1(Lines))



def part_2(Games):
    maxes = []
    sum=0
    for game in Games:
        maxG=maxB=maxR=0
        gameId=int(game.split(":")[0].split("Game")[1].strip() )
        draws =game.split(":")[1].split(';')
        for draw in draws:
            reds = blues = greens = 0
            for colour in draw.split(","):
                if 'blue' in colour:
                    blues = int(colour.split('blue')[0].strip())
                    if blues > maxB: maxB = blues
                if 'red' in colour:
                    reds = int(colour.split('red')[0].strip())
                    if reds > maxR: maxR=reds
                if 'green' in colour:
                    greens = int(colour.split('green')[0].strip())
                    if greens > maxG: maxG=greens
        maxes.append((gameId,maxR,maxG,maxB))
        sum+= maxR*maxG*maxB
    return sum

print(part_2(Lines))

File: 2023/Day2/day2.scala
import scala.io.Source

object Main {

  def main(args: Array[String]): Unit = {

    val Lines: List[String] =
      if   (args.length > 0) Source.fromFile(args(0)).getLines.toList
      else Nil

    if (Lines.length == 0){
      println("Please enter file name")
      scala.sys.exit(1)
      }
    //Lines.foreach(println)

    // Convert each draw into 3 numbers, and then check each of these three
    // for a condition. We can use a tuple: (R,G,B)
    val Games: List[Array[String]] = for(line<-Lines) yield line.split(":")(1).split(";")
    val tupTriplets =
      for (draws <- Games) yield for(draw<-draws) yield {
      val colours = draw.split(",")

      val red   = for(colour<-colours; if colour.contains("red");   letter<-colour; if letter.isDigit) yield letter
      val blue  = for(colour<-colours; if colour.contains("blue");  letter<-colour; if letter.isDigit) yield letter
      val green = for(colour<-colours; if colour.contains("green"); letter<-colour; if letter.isDigit) yield letter

      val redz   = if(red.isEmpty)   0 else red.mkString.toInt
      val bluez  = if(blue.isEmpty)  0 else blue.mkString.toInt
      val greenz = if(green.isEmpty) 0 else green.mkString.toInt
      (redz,bluez,greenz)
    }

    // part 1
    val rMax = 12; val bMax = 14; val gMax = 13
    def possGame(tupArray: Array[(Int,Int,Int)]) : Boolean =
      tupArray.forall(s=> s._1 <= rMax && s._2 <= bMax && s._3 <=gMax)

    val possibleAndIndex = tupTriplets.zipWithIndex.filter(s => possGame(s._1) )
    println(possibleAndIndex.map(_._2+1).sum)

    // part 2
    def maxPower(tupArray: Array[(Int,Int,Int)]) : Int =
      tupArray.map(_._1).max*tupArray.map(_._2).max*tupArray.map(_._3).max
    println(tupTriplets.map(maxPower(_)).sum )


    }

}
